### PR TITLE
MAYA-127393 allow USD ref to be relative

### PR DESCRIPTION
**Pull request by pierrebai-adsk**
- Refactor the Python helper to handle the relative-path UI in file dialogs.
- Make it more generic by abstracting to which file the path will be relative to, the label of the UI and the option var used.
- Add makePathRelativeTo function to UsdMayaUtilFileSystem to make any file path relative to a given directory.
- Add requireUsdPathsRelativeToEditTargetLayer function to verify if USD ref paths should be relative.
- Make the USD context menu use the new relative ref UI and functions.
- Refactor its code a bit to make reading the generated scripts easier.

### CODE REVIEW

File: lib/mayaUsd/utils/utilFileSystem.cpp [hunk 1]
```diff
@@ -117,27 +117,42 @@ std::string UsdMayaUtilFileSystem::getMayaSceneFileDir()
...
+    if (relativeToDir.empty()) {
+        return std::make_pair(fileName, true);
     }
 
-    ghc::filesystem::path relativePath = absolutePath.lexically_relative(basePath);
+    ghc::filesystem::path relativePath = absolutePath.lexically_relative(relativeToDir);
 
     if (relativePath.empty()) {
+        return std::make_pair(fileName, false);
+        ;
```
Comment (vlasovi) [inline lib/mayaUsd/utils/utilFileSystem.cpp:138]: Minor typo - this line can be removed.

**Review by vlasovi — DISMISSED** [review]

File: plugin/adsk/scripts/mayaUsd_USDRootFileRelative.py [hunk 2]
```diff
@@ -3,18 +3,50 @@
...
 
-class usdRootFileRelative(object):
-    fileNameEditField = None
+class usdFileRelative(object):
+    '''
+    Helper class to create the UI for load/save dialog boxes that need to make the
+    selected file name optionally relative.
 
-    kMakePathRelativeCheckBox = 'MakePathRelative'
+    The caller must tell each function to what the file should be made relative.
```
Comment (seando-adsk) [inline plugin/adsk/scripts/mayaUsd_USDRootFileRelative.py:11]:
```suggestion
    The caller must tell each function to what the file should be made relative to.
```

File: plugin/adsk/scripts/mayaUsd_USDRootFileRelative.py [hunk 3]
```diff
@@ -3,18 +3,50 @@
...
+    The 'what' is used to select the correct UI element, UI label and to read
+    and write the correct option var.
```
Comment (seando-adsk) [inline plugin/adsk/scripts/mayaUsd_USDRootFileRelative.py:13]: Nice trick for re-using all this code.

File: plugin/adsk/scripts/mayaUsd_USDRootFileRelative.py [hunk 4]
```diff
@@ -32,32 +64,114 @@ def uiCreate(cls, parentLayout):
...
+        cmds.checkBox(cls.kMakePathRelativeCheckBox + relativeToWhat, label=kMakePathRelativeStr, ann=kMakePathRelativeAnnStr)
 
     @classmethod
-    def uiInit(cls, parentLayout, filterType):
+    def uiInit(cls, parentLayout, canBeRelative, relativeToWhat):
+        """
+        Helper method to initialize the UI layout for the file relative actions.
+
+        Input parentLayout arg is expected to the a scroll layout into which controls
+        can be added.
```
Comment (seando-adsk) [inline plugin/adsk/scripts/mayaUsd_USDRootFileRelative.py:82]: Sorry I realize I made a grammar mistake when I added this. Can you fix it. 
```suggestion
        Input parentLayout arg is expected to be a scroll layout into which controls
        can be added.
```

File: plugin/adsk/scripts/mayaUsd_USDRootFileRelative.py [hunk 5]
```diff
@@ -32,32 +64,114 @@ def uiCreate(cls, parentLayout):
...
+        usdFileRelative.uiCreate(parentLayout, cls.kRelativeToWhat)
+
+    @classmethod
+    def uiInit(cls, parentLayout, filterType):
+        '''
+        Note: the function takes an unused filterType argument to be compatible
+              with the dialog2 command API.
+        '''
+        # If there is no Maya scene file saved, then the checkbox and label should be disabled.
+        print(usdFileRelative.getRelativeFilePathRoot())
```
Comment (seando-adsk) [inline plugin/adsk/scripts/mayaUsd_USDRootFileRelative.py:167]: print should be removed?

**Review by seando-adsk — CHANGES_REQUESTED** [review]
Nice work overall, just a few things to fixup.

File: lib/mayaUsd/ufe/UsdContextOps.cpp [hunk 6]
```diff
@@ -473,58 +475,107 @@ class ToggleInstanceableStateCommand : public Ufe::UndoableCommand
...
+        const char* script = R"mel(
+        global proc string SelectUSDFileForAddReference()
+        {
+            string $result[] = `fileDialog2
+                -fileMode 1
+                -caption "Add Reference to USD Prim"
+                -fileFilter "USD Files (%s);;%s"
+                -optionsUICreate addUSDReferenceCreateUi
+                -optionsUIInit addUSDReferenceInitUi
+                -selectionChanged addUSDReferenceSelectionChanged
```
[the remark this answers is not in the document]
Comment (pierrebai-adsk) [reply lib/mayaUsd/ufe/UsdContextOps.cpp:541]: OK, I thought maybe in the future if we show a preview of the relative path it could be useful, but we can add it later if needed.

File: plugin/adsk/scripts/mayaUsd_USDRootFileRelative.py (hunk 2 above)
Comment (pierrebai-adsk) [reply plugin/adsk/scripts/mayaUsd_USDRootFileRelative.py:11]: I'll do the fix in the same commit as the rest to avoid having dozens of commits.

**Review by vlasovi — APPROVED** [review]

**Review by vlasovi — APPROVED** [review]

File: plugin/adsk/scripts/mayaUSDRegisterStrings.py [hunk 7]
```diff
@@ -36,6 +36,6 @@ def mayaUSDRegisterStrings():
     register("kLoadUSDFile", "Load USD File")
     register("kFileOptions", "File Options")
     register("kMakePathRelativeToSceneFile", "Make Path Relative to Scene File")
-    register("kMakePathRelativeToSceneFileAnn", "If enabled, path will be relative to your Maya scene file. If this option is disabled, there is no Maya scene file and the path will be absolute. Save your Maya scene file to disk to make this option available.")
+    register("kMakePathRelativeToSceneFileAnn", "If enabled, path will be relative to your Maya scene file.\nIf this option is disabled, there is no Maya scene file and the path will be absolute.\nSave your Maya scene file to disk to make this option available.")
```
Comment (seando-adsk) [inline plugin/adsk/scripts/mayaUSDRegisterStrings.py:39]: Why this change? Were you asked by documentation to make it? I originally added this string exactly as given to me by documentation and QA verified it.

Comment (pierrebai-adsk) [reply plugin/adsk/scripts/mayaUSDRegisterStrings.py:39]: The only change is to split it on two line. It seemed hard to read when all on a single very long line: the tooltip was almost as wide as my monitor. I did not think the line break, splitting each sentence on its own line would be important.

**Review by seando-adsk — APPROVED** [review]

File: plugin/adsk/scripts/mayaUSDRegisterStrings.py (hunk 7 above)
Comment (pierrebai-adsk) [reply plugin/adsk/scripts/mayaUSDRegisterStrings.py:39]: Confirming: we asked the designers and they are OK with splitting the lines.
